drivers/video/video_stm32_venc.c

```diff
@@ -0,0 +1,899 @@
+
+	do {
+		irq_stats = sys_read32(config->reg + BASE_HEncIRQ);
+		/* get the number of completed slices from ASIC registers. */
+		if (slicesReady != NULL && *slicesReady > prevSlicesReady) {
+			*slicesReady = FIELD_GET(NUM_SLICES_READY_MASK,
+						 sys_read32(config->reg + BASE_HEncControl7));
+		}
+
+		LOG_DBG("IRQ stat = %08x", irq_stats);
+
+		uint32_t hw_handshake_status = IS_BIT_SET(
+			sys_read32(config->reg + BASE_HEncInstantInput), LOW_LATENCY_HW_ITF_EN);
+
+		/* ignore the irq status of input line buffer in hw handshake mode */
+		if ((irq_stats == ASIC_STATUS_LINE_BUFFER_DONE) && (hw_handshake_status != 0UL)) {
+			sys_write32(ASIC_STATUS_FUSE, config->reg + BASE_HEncIRQ);
+			continue;
+		}
+
+		if ((irq_stats & ASIC_STATUS_ALL) != 0UL) {
+			/* clear IRQ and slice ready status */
+			uint32_t clr_stats;
+
+			irq_stats &= ~(ASIC_STATUS_SLICE_READY | ASIC_IRQ_LINE);
+
+			if (clrByWrite1 != 0UL) {
+				clr_stats = ASIC_STATUS_SLICE_READY | ASIC_IRQ_LINE;
+			} else {
+				clr_stats = irq_stats;
+			}
+
+			sys_write32(clr_stats, config->reg + BASE_HEncIRQ);
+			ret = EWL_OK;
+			break;
+		}
+
+		if (slicesReady != NULL && *slicesReady > prevSlicesReady) {
+			ret = EWL_OK;
+			break;
+		}
+
+	} while (!sys_timepoint_expired(timeout));
+
+	LOG_DBG("encoding = %d ms", k_ticks_to_ms_ceil32(sys_clock_tick_get_32() - start));
+
+	if (slicesReady != NULL) {
+		LOG_DBG("slicesReady = %d", *slicesReady);
+	}
+
+	return ret;
+}
+
+void EWLassert(bool expr, const char *str_expr, const char *file, unsigned int line)
+{
+	__ASSERT(expr, "ASSERTION FAIL [%s] @ %s:%d", str_expr, file, line);
+}
+
+/* Set CONFIG_VC8000NANOE_LOG_LEVEL_DBG to enable library tracing */
+void EWLtrace(const char *s)
+{
+	printk("%s\n", s);
+}
+
+void EWLtraceparam(const char *fmt, const char *param, unsigned int val)
+{
+	printk(fmt, param, val);
+}
+
+static int stm32_venc_enable_clock(const struct device *dev)
+{
+	const struct stm32_venc_config *config = dev->config;
+	const struct device *clk = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+
+	if (!device_is_ready(clk)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	if (clock_control_on(clk, (clock_control_subsys_t)&config->pclken) != 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int stm32_venc_set_fmt(const struct device *dev, struct video_format *fmt)
+{
+	struct stm32_venc_data *data = dev->data;
+
+	if (fmt->type == VIDEO_BUF_TYPE_INPUT) {
+		if ((fmt->pixelformat != VIDEO_PIX_FMT_NV12) &&
+		    (fmt->pixelformat != VIDEO_PIX_FMT_RGB565)) {
+			LOG_ERR("invalid input pixel format");
+			return -EINVAL;
+		}
+
+		fmt->pitch = fmt->width * video_bits_per_pixel(fmt->pixelformat) / BITS_PER_BYTE;
+		data->in_fmt = *fmt;
+	} else {
+		if (fmt->pixelformat != VIDEO_PIX_FMT_H264) {
+			LOG_ERR("invalid output pixel format");
+			return -EINVAL;
+		}
+
+		fmt->size = fmt->width * fmt->height / VENC_ESTIMATED_COMPRESSION_RATIO;
```

Comment:
A compression ratio would be comparing the input size (`fmt->pitch` rather than `fmt->width`) to the output size (in H.264), I suppose here all pixel formats (RGB565, NV12) are first converted to a same format, then compressed?

Otherwise the output size (bytesused) and compression ratio would be different for NV12 and RGB565?

Not blocking and I was just curious about it.
Thanks!

Reply:
Very good to know. This makes sense the way it is and having fine estimation are not blockers of course.
Thank you.